### ISSUE TITLE
Move Linux user data to XDG directories

### ISSUE THREAD

**Issue by Montandalar** [minetest/minetest#15382]
### Problem

Minetest put* (and Luanti, 5.10, will put) a `.minetest` directory in your home directory. In the Linux world, this is considered an outdated practice. With the average desktop user installing hundreds of software packages, if each one did this, it quickly adds up to a huge number of subdirectories.

With the rename to Luanti, there is now an opportunity to address this at the same time, while still remaining backwards-compatible by using priorities.

*past tense

### Solutions

The [XDG Base Directory Specifications](https://specifications.freedesktop.org/basedir-spec/latest/) specify where user data for applications software should go. In the case of Luanti, this would mean two new paths:

`~/.minetest/minetest.conf` -> `~/.config/luanti/luanti.conf`
`~/.minetest/` -> `~/.local/share/luanti/`
 (the rest of the directory with its subdirectories)

Probably `clientmods/mods.conf` would be moved to `~/.config` as well.

These paths are only a guide, applications are supposed to follow the environment variables specified by the XDG spec properly if those are defined.

### Alternatives

Leave the name `.minetest` in the home directory in perpetuity for backwards compatibility, confusing users.

Rename the directory to `.luanti` for new installations, checking for `.minetest` for some period (or perpetuity), and using that instead, or renaming it.

### Additional context

The move may be delayed for some fixed period like a year for the purpose of letting the new name settle in, but should not stay as-is in perpetuity.

A drop in backwards-compatibility to the old path should not happen before version 6.0.0 of the program.

**Comment by GoodClover** [comment]
It would be worth adding a warning or error for if both exist, otherwise I imagine there'll be some very frustrating confusion if you somehow end up with both unknowingly.

**Comment by jedenastka** [comment]
Please do not use `~/.local/share`, which is equivalent to `/usr/share`, and rather use `~/.local/state`, which is equivalent to `/var/lib`.

Also, those are not really hard paths, and rather default values for environment variables:
- `~/.config`: `XDG_CONFIG_HOME`
- `~/.local/state`: `XDG_STATE_HOME`
- `~/.cache`: `XDG_CACHE_HOME`

This seems to be already taken into account for the current cache location.

**Comment by Montandalar** [comment]
> Please do not use `~/.local/share`, which is equivalent to `/usr/share`, and rather use `~/.local/state`, which is equivalent to `/var/lib`.

But these files do live in /usr/share/luanti when installed as a system-wide package. Nor does Luanti use /var/lib at all.

I also checked my own state directory, and there is very little there except some files that track small details about some KDE apps, with information about the size of my my desktop across two screens. The only thing I can think off off-hand would be the current main menu tab. The XDG specification indicates logs might live there, but I think hiding debug.txt away there would be a worse option than leaving it where it would be expected with the DATA directory.


> Also, those are not really hard paths, and rather default values for environment variables:
> 
>     * `~/.config`: `XDG_CONFIG_HOME`
> 
>     * `~/.local/state`: `XDG_STATE_HOME`
> 
>     * `~/.cache`: `XDG_CACHE_HOME`
> 
> 
> This seems to be already taken into account for the current cache location.

Thank you for pointing it out, but I think leaving the default paths helps the clarity of OP.

**Comment by jedenastka** [comment]
> But these files do live in /usr/share/luanti when installed as a system-wide package. Nor does Luanti use /var/lib at all.

Yes, but say, `PREFIX/share/luanti/textures` (with `PREFIX` being `/usr` by default) is **not** equivalent to `~/.minetest/textures`. `PREFIX/share/luanti/textures` is where the preinstalled textures live, `~/.minetest/textures` is where textures are *installed* to.

(Admittedly, this should be `PREFIX/lib`, as it's program data, rather than actually shared data, but that is another story, and not something I want to discuss here.)

You would not want a program to modify it's own `/usr/share`, nor would you want a user to meddle with it's contents. Meanwhile `/var/lib` is fair game for both the program and the user. Yes, Minetest doesn't use that, because it has no need for system-wide variable data, however, it does have a need for user-wide variable data.

Moreover, Luanti can be reasonably built with `PREFIX` equal to `~/.local`. This causes both of these very separate directories to be conflated. Not a great idea.

> I also checked my own state directory, and there is very little there except some files that track small details about some KDE apps, with information about the size of my my desktop across two screens.

Yes. That is because most people do this wrong, perhaps due to `XDG_STATE_HOME` being a "recent" (2021) addition, and also honestly documented in a way that is not giving what it was intended to be used justice.

> Thank you for pointing it out, but I think leaving the default paths helps the clarity of OP.

I don't think so. If this was implemented as it is stated in the original post, it would be implemented wrong. We should be clear on what we want in bug reports.

**Comment by Montandalar** [comment]
### 1. On the separation of directories

> Yes, but say, `PREFIX/share/luanti/textures` (with `PREFIX` being `/usr` by default) is **not** equivalent to `~/.minetest/textures`. `PREFIX/share/luanti/textures` is where the preinstalled textures live, `~/.minetest/textures` is where textures are _installed_ to.
> 
> (Admittedly, this should be `PREFIX/lib`, as it's program data, rather than actually shared data, but that is another story, and not something I want to discuss here.)
> 
> You would not want a program to modify it's own `/usr/share`, nor would you want a user to meddle with it's contents. Meanwhile `/var/lib` is fair game for both the program and the user. Yes, Minetest doesn't use that, because it has no need for system-wide variable data, however, it does have a need for user-wide variable data.

Mods and games can be installed system-wide and are through Debian packages (not that Debian's mod packages are good, but they) - before Minetest Game ceased to be shipped with the engine, there was even that game pre-installed with the package in a system-wide location. This would work well for deploying Luanti with a pre-made game to a lot of school laptops, for instance. Mods, games and texturepacks aren't meant to modify their own content either.

System-wide, it seems like worlds would live inside of /var/lib. But of course, this isn't a path used or checked, and I don't think system-wide worlds are desired.

I do not think it contradicts you to say: Luanti has a clear precedent of system-installed content (e.g. `/usr/local/share/luanti/textures/base/pack`) and user content (e.g. `~/.minetest/textures`) being separate.

### 2. On the semantics attached to directories

> Moreover, Luanti can be reasonably built with `PREFIX` equal to `~/.local`. This causes both of these very separate directories to be conflated. Not a great idea.

>>    I also checked my own state directory, and there is very little there except some files that track small details about some KDE apps, with information about the size of my my desktop across two screens.

>Yes. That is because most people do this wrong, perhaps due to XDG_STATE_HOME being a "recent" (2021) addition, and also honestly documented in a way that is not giving what it was intended to be used justice.

Can we get some links and references about intent here? Because it sounds like an information propagation problem. If it's not written into the XDG specification, but found on a mailing list, it's likely only a few people have read it. If this were Microsoft Windows, a small product team would have ownership and make a decision and write documentation like [Store and retrieve settings and other app data](https://learn.microsoft.com/en-us/windows/apps/design/app-settings/store-and-retrieve-app-data), but Linux is so decentralised (Mind you, to find the Microsoft article, I refined my search engine to use `site:microsoft.com`; but I can't do that across the hundreds of domains of Linux-associated projects to more easily find primary sources).

I get that doesn't help that the state directory is newer, and "people do it wrong", but of course that we want to do this right if we're going to. If I go off of the specification as it is written:

> The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain: 

I am inclined to say that Luanti doesn't have anything except logs that fits there. The content (mods, clientmods games and texture packs) would live in `$XDG_DATA_HOME`. The program itself would live in ~/.local/bin. The cache has its own directory. Surely worlds would also live in `$XDG_DATA_HOME`. Maybe I'm just underinformed or misinformed

 -- It just sounds to me like the contents of `$XDG_CACHE_HOME` and `$XDG_STATE_HOME` could be deleted without the user losing data about their own created files, and after looking through the contents of both of my own I can say that looks correct for what I currently have installed, but not if my worlds were in `$XDG_DATA_HOME`.


### 3. Given the two above

So while in section (1) agree that there is a separation of Luanti "system" files from user files, and that the user may have a prefix inside of `.local`, putting the program in e.g.  `~/.local/bin` and files in `$XDG_DATA_HOME`, in section (2) I did not manage to find a good source on the semantic meaning of `$XDG_STATE_HOME` being the correct solution to this separation of concerns, rather it sounds more temporary of a location for files. I hope you can easily provide this for my benefit, who is probably just not very well-informed.

### Addendum
> If this was implemented as it is stated in the original post [ed: bare directory names], it would be implemented wrong. We should be clear on what we want in bug reports.

Fair point, I have edited OP to point out the use of variables, making explicit rather than the implicit nature of referring to XDG in hopes of following that properly.

**Comment by rollerozxa** [comment]
Yes, `$XDG_DATA_HOME` is the correct place Luanti should be putting persistent user data into. I don't know where this idea comes from that it should be in `$XDG_STATE_HOME`, which is intended to be more persistent than `$XDG_CACHE_HOME` but still ephemeral. You should be able to safely delete `$XDG_STATE_HOME` and restart without causing data loss, putting important data such as users' worlds there is ***not*** something you should do.

**Comment by Sheriff-Unit-3** [comment]
As a Linux user this would be nice to see by Luanti v6.0.0 at latest, but earlier would be even better.
